pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,600 @@
+	m.usedBy[string(pod)] = from
+	return nil
+}
+
+// Release releases the user namespace allocated to the specified pod.
+func (m *usernsManager) Release(podUID types.UID) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		return
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	m.releaseWithLock(podUID)
+}
+
+func (m *usernsManager) releaseWithLock(podUID types.UID) {
+	pod := string(podUID)
+	v, ok := m.usedBy[pod]
+	if !ok {
+		klog.V(5).InfoS("pod user namespace allocation not present", "podUID", podUID)
+		return
+	}
+	delete(m.usedBy, pod)
+
+	klog.V(5).InfoS("releasing pod user namespace allocation", "podUID", podUID)
+	m.numAllocated--
+	m.removed++
+
+	if m.removed%mapReInitializeThreshold == 0 {
+		n := make(map[string]uint32)
+		for k, v := range m.usedBy {
+			n[k] = v
+		}
+		m.usedBy = n
+		m.removed = 0
+	}
+	m.used.clear(v / userNsLength)
+}
+
+func (m *usernsManager) parseUserNsFileAndRecord(pod types.UID, content []byte) (userNs userNamespace, err error) {
+	if err = json.Unmarshal([]byte(content), &userNs); err != nil {
+		err = fmt.Errorf("can't parse file: %w", err)
+		return
+	}
+
+	if len(userNs.UIDMappings) != 1 {
+		err = fmt.Errorf("invalid user namespace configuration: no more than one mapping allowed.")
+		return
+	}
+
+	if len(userNs.UIDMappings) != len(userNs.GIDMappings) {
+		err = fmt.Errorf("invalid user namespace configuration: GID and UID mappings should be identical.")
+		return
+	}
+
+	if userNs.UIDMappings[0] != userNs.GIDMappings[0] {
+		err = fmt.Errorf("invalid user namespace configuration: GID and UID mapping should be identical")
+		return
+	}
+
+	// We don't produce configs without root mapped and some runtimes assume it is mapped.
+	// Validate the file has something we produced and can digest.
+	if userNs.UIDMappings[0].ContainerId != 0 {
+		err = fmt.Errorf("invalid user namespace configuration: UID 0 must be mapped")
+		return
+	}
+
+	if userNs.GIDMappings[0].ContainerId != 0 {
+		err = fmt.Errorf("invalid user namespace configuration: GID 0 must be mapped")
+		return
+	}
+
+	hostId := userNs.UIDMappings[0].HostId
+	length := userNs.UIDMappings[0].Length
+
+	err = m.record(pod, hostId, length)
+	return
+}
+
+func (m *usernsManager) createUserNs(pod *v1.Pod) (userNs userNamespace, err error) {
+	firstID, length, err := m.allocateOne(string(pod.UID))
+	if err != nil {
+		return
+	}
+
+	defer func() {
+		if err != nil {
+			m.Release(pod.UID)
+		}
+	}()
+
+	userNs = userNamespace{
+		UIDMappings: []idMapping{
+			{
+				ContainerId: 0,
+				HostId:      firstID,
+				Length:      length,
+			},
+		},
+		GIDMappings: []idMapping{
+			{
+				ContainerId: 0,
+				HostId:      firstID,
+				Length:      length,
+			},
+		},
+	}
+
+	return userNs, m.writeMappingsToFile(pod.UID, userNs)
+}
+
+// GetUserNamespaceMappings returns the configuration for the sandbox user namespace
+func (m *usernsManager) GetUserNamespaceMappings(pod *v1.Pod) (*runtimeapi.UserNamespace, error) {
```

Comment:
the implementation of this method is actually `GetOrCreateUserNamespaceMappings`, can we update to reflect that its actually mutating state?

Reply:
Changed to that name too